core/src/main/java/io/questdb/griffin/engine/functions/str/SplitPartFunctionFactory.java

```diff
@@ -0,0 +1,210 @@
+                return StrConstant.NULL;
+            } else if (index == 0) {
+                return StrConstant.EMPTY;
+            } else {
+                return new SplitPartConstIndexFunction(strFunc, delimiterFunc, index);
+            }
+        }
+        return new SplitPartFunction(strFunc, delimiterFunc, indexFunc);
+    }
+
+    private static class SplitPartFunction extends AbstractSplitPartFunction implements TernaryFunction {
+        private final Function strFunc;
+        private final Function delimiterFunc;
+        private final Function indexFunc;
+
+        public SplitPartFunction(Function strFunc, Function delimiterFunc, Function indexFunc) {
+            super(strFunc, delimiterFunc);
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+            this.indexFunc = indexFunc;
+        }
+
+        @Override
+        int getIndex(Record rec) {
+            return indexFunc.getInt(rec);
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getCenter() {
+            return delimiterFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return indexFunc;
+        }
+    }
+
+    private static class SplitPartConstIndexFunction extends AbstractSplitPartFunction implements BinaryFunction {
+        private final Function strFunc;
+        private final Function delimiterFunc;
+        private final int index;
+
+        public SplitPartConstIndexFunction(Function strFunc, Function delimiterFunc, int index) {
+            super(strFunc, delimiterFunc);
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+            this.index = index;
+        }
+
+        @Override
+        int getIndex(Record rec) {
+            return index;
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return delimiterFunc;
+        }
+    }
+
+    private static abstract class AbstractSplitPartFunction extends StrFunction {
+        private final StringSink sink = new StringSink();
+        private final StringSink sinkB = new StringSink();
+
+        private final Function strFunc;
+        private final Function delimiterFunc;
+
+        public AbstractSplitPartFunction(Function strFunc, Function delimiterFunc) {
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+        }
+
+        abstract int getIndex(Record rec);
+
+        @Override
+        public CharSequence getStr(Record rec) {
+            return getStr0(rec, sink);
+        }
+
+        @Override
+        public CharSequence getStrB(Record rec) {
+            return getStr0(rec, sinkB);
+        }
+
+        @Override
+        public void getStr(Record rec, CharSink sink) {
+            getStr0(rec, sink);
```

Comment:
This method shouldn't call `sink.clear()` - its contract is to append the chars to the provided sink, without clearing it. To achieve that, simply call `sink.clear()`/`sinkB.clear()` in `getStr()`/`getStrB()` calls.

Reply:
It won't, `getStr0` will only call `sink.clear()` if the `CharSink` implements `Mutable`. `StringSink` implements `Mutable`, while `CharSink` doesn't extend `Mutable`, so it will call `sink.clear()` if its called from `getStr(Record, StringSink)`/`getStrB(Record, StringSink)`, but not `getStr(Record, CharSink)`.

See my comment above (https://github.com/questdb/questdb/pull/2644#discussion_r996312515) for more information.

Reply:
Ah, I see, it shouldn't clear it even if the `CharSink`implements `Mutable`. I'll add a `boolean` `clear` parameter to `getStr0(Record, CharSink)` then

Reply:
There is nothing preventing `CharSink` implementations to also implement `Mutable` (for instance, see `MutableCharSink`).

Reply:
I've added a `boolean` `clearSink` parameter to `getStr0()`. Now it only clears the sink if `clearSink` is `true` and `sink` is an instance of `Mutable`.

IMO this is cleaner than duplicating the logic between `getStr(Record, StringSink)`/`getStrB(Record, StringSink)` and `getStr(Record, CharSink)`.